File: python/354_俄罗斯套娃信封问题.py
import bisect
class solution354:
    def maxEnvelopes(self,envelopes):
        n=len(envelopes)
        if n<=0:
            return n
        envelopes.sort(key=lambda x:(x[0],-x[1]))
        #f=[]
        #f.append(envelopes[0][1])
        f=[envelopes[0][1]]
        for i in range(1,n):
            #if (num:=envelopes[i][1])>f[-1]: #python3.8 以上的海象运算
            num=envelopes[i][1]
            if num > f[-1]:
                f.append(num)
            else:
                index=bisect.bisect_left(f,num)
                f[index]=num
        return len(f)
if __name__ == '__main__':
    #envelopes=[[5,4],[6,4],[6,7],[2,3]]
    envelopes=eval(input("请输入如：[[5,4],[6,4],[6,7],[2,3]]的二维列表"))
    s354=solution354()
    print('\n')
    print(s354.maxEnvelopes(envelopes))

File: c++/354_俄罗斯套娃信封问题.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
class solution_673 {
	public:
		int maxEnvelopes(vector<vector<int>>& envelopes) {
			int n = envelopes.size();
			if (n == 0) return 0;
			sort(envelopes.begin(), envelopes.end(), [](vector<int>& a, vector<int>& b) {return a[0] < b[0] || (a[0] == b[0] && a[1] > b[1]);});
			vector<int> dp = { envelopes[0][1] };
			for (int i = 0; i < n; ++i) {
				//if (envelopes[i][0] == envelopes[i - 1][0]||envelopes[i][1]==envelopes[i-1][1]) continue;
				if (envelopes[i][1] > dp.back()) {
					dp.push_back(envelopes[i][1]);
				}
				else {
					int p=lower_bound(dp.begin(),dp.end(),envelopes[i][1])-dp.begin();
					dp[p]=envelopes[i][1];
				}
			}
			return dp.size();

	}
};
int main() {
	vector<vector<int>> input;
	int a, b;
	while (cin >> a&&cin>>b) {
		input.push_back({ a,b });
		if (cin.get() == '\n') break;
	}
	solution_673 s_673;
	cout<<s_673.maxEnvelopes(input);

}